lib/grandpa/grandpa_test.go

```diff
@@ -1279,3 +1280,127 @@ func TestFinalRoundGaugeMetric(t *testing.T) {
 	gauge := ethmetrics.GetOrRegisterGauge(finalityGrandpaRoundMetrics, nil)
 	require.Equal(t, gauge.Value(), int64(180))
 }
+
+func TestGrandpaServiceCreateJustification_ShouldCountEquivocatoryVotes(t *testing.T) {
+	t.Parallel()
+
+	// setup granpda service
+	gs, st := newTestService(t)
+	addBlocksToState(t, st.Block, 10)
+
+	// create a new fake block to fake authorites vote on
+	previousHash := st.Block.BestBlockHash()
+	previusHead, err := st.Block.BestBlockHeader()
+	require.NoError(t, err)
+
+	bfcNumber := int(previusHead.Number.Int64() + 1)
+
+	d, err := types.NewBabePrimaryPreDigest(0, uint64(bfcNumber), [32]byte{}, [64]byte{}).ToPreRuntimeDigest()
+	require.NoError(t, err)
+	require.NotNil(t, d)
+	digest := types.NewDigest()
+	_ = digest.Add(*d)
+
+	bfcBlock := &types.Block{
+		Header: types.Header{
+			ParentHash: previousHash,
+			Number:     big.NewInt(int64(bfcNumber)),
+			StateRoot:  trie.EmptyHash,
+			Digest:     digest,
+		},
+		Body: types.Body{},
+	}
+
+	bfcHash := bfcBlock.Header.Hash()
+	err = st.Block.AddBlockWithArrivalTime(bfcBlock, time.Now())
+	require.Nil(t, err)
+
+	// create fake authorities
+	fakeAuthorities := make([]*ed25519.Keypair, 0)
+	ed25519Keyring, _ := keystore.NewEd25519Keyring()
+
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Alice().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Bob().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Charlie().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Dave().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Eve().(*ed25519.Keypair))
+
+	equivocatories := make(map[ed25519.PublicKeyBytes][]*types.GrandpaSignedVote)
+	prevotes := &sync.Map{}
+
+	// voting on
+	for i, v := range fakeAuthorities {
+		vote := &SignedVote{
+			AuthorityID: v.Public().(*ed25519.PublicKey).AsBytes(),
+			Signature:   [64]byte{},
+			Vote: types.GrandpaVote{
+				Hash:   bfcHash,
+				Number: uint32(bfcNumber),
+			},
+		}
+
+		prevotes.Store(i, vote)
+
+		// lets say authority 1 and 3 have equivocatory votes
+		if i == 1 || i == 3 {
+			equivocatories[v.Public().(*ed25519.PublicKey).AsBytes()] = []*types.GrandpaSignedVote{
+				{
+					Vote: types.GrandpaVote{
+						Hash:   common.Hash{},
+						Number: 0,
+					},
+					Signature:   [64]byte{},
+					AuthorityID: v.Public().(*ed25519.PublicKey).AsBytes(),
+				},
+			}
```

Comment:
nit can you split this please

```suggestion
			key := v.Public().(*ed25519.PublicKey).AsBytes()
			value := []*types.GrandpaSignedVote{
				{
					Vote: types.GrandpaVote{
						Hash:   common.Hash{},
						Number: 0,
					},
					Signature:   [64]byte{},
					AuthorityID: v.Public().(*ed25519.PublicKey).AsBytes(),
				},
			}
			equivocatories[key] = value
```